src/workers/parser/mapBindings.js

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */
+
+// @flow
+
+import { parseScript } from "./utils/ast";
+import { buildScopeList } from "./getScopes";
+import generate from "@babel/generator";
+import * as t from "@babel/types";
+import remove from "lodash/remove";
+
+import type { ColumnPosition } from "../../types";
+
+// translates new bindings `var a = 3` into `self.a = 3`
+// and existing bindings `var a = 3` into `a = 3` for re-assignments
+function globalizeAssignment(node, bindings) {
+  return node.declarations.map(node => {
+    const identifier = bindings.includes(node.id.name)
+      ? node.id
+      : t.memberExpression(t.identifier("self"), node.id);
+
+    return t.expressionStatement(
+      t.assignmentExpression("=", identifier, node.init)
+    );
+  });
+}
+
+function isTopLevel(ancestors) {
+  return ancestors.filter(ancestor => ancestor.key == "body").length == 1;
+}
+
+function replaceParent(ancestors, newNodes) {
+  const parent = ancestors[ancestors.length - 1];
+  const lastAncestor = ancestors[ancestors.length - 1];
+  const { index } = lastAncestor;
+  return parent.node[parent.key].splice(index, 1, ...newNodes);
+}
+
+function hasDestructuring(node) {
+  return node.declarations.some(declaration => t.isPattern(declaration.id));
+}
+
+function alreadyDefined(node, bindings) {
+  const names = node.declarations.map(dec => dec.id.name);
+  return names.some(name => bindings.includes(name));
+}
+
+export default function mapExpressionBindings(
```

Comment:
Should we skip entirely if a `with` statement is found? I think probably? For example:
```
var obj = { a: 5 };
with(obj) {
  var a = 4;
  a = 6;
}
```
if the `var a = 4` becomes `self.a = 4` then `a = 6` is suddenly modifying the value on the object instead.

Reply:
If we're entirely bailing out there, it also makes be wonder if we'd be better bailing entirely when destructuring is used. I worry that it might be even more confusing to users if part of their executed snippet behaves one way, and part behaves another way.

Reply:
> should we skip entirely if a `with` is found
yep.

> bailing entirely
I agree. better to be safe than sorry